anndata/_config.py

```diff
@@ -0,0 +1,332 @@
+from __future__ import annotations
+
+import os
+import warnings
+from collections.abc import Iterable
+from contextlib import contextmanager
+from functools import wraps
+from inspect import Parameter, signature
+from typing import TYPE_CHECKING, Any, NamedTuple, TypeVar
+
+if TYPE_CHECKING:
+    from collections.abc import Callable, Sequence
+
+T = TypeVar("T")
+
+
+class DeprecatedOption(NamedTuple):
+    option: str
+    message: str | None
+    removal_version: str | None
+
+
+# TODO: inherit from Generic[T] as well after python 3.9 is no longer supported
+class RegisteredOption(NamedTuple):
+    option: str
+    default_value: T
+    doc: str
+    validator: Callable[[T], None] | None
+
+
+def update_override_function(register: Callable):
+    """This function updates the keyword arguments, docstring, and annotations of the `SettingsManager.override` function as the `SettingsManager.register` method is called.
+
+    Parameters
+    ----------
+    register
+        This is a reference to the register function of `SettingsManager`
+
+    Returns
+    -------
+    Callable
+        The wrapped `SettingsManager.register` function.
+    """
+
+    @wraps(register)
+    def _impl(
+        self,
+        option: str,
+        default_value: object,
+        doc: str,
+        validator: Callable[[T], None],
+    ):
+        method_output = register(self, option, default_value, doc, validator)
+        # Update annotations for type checking.
+        self.override.__annotations__[option] = type(default_value)
+        # __signature__ needs to be updated for tab autocompletion in IPython.
+        # See https://github.com/ipython/ipython/issues/11624 for inspiration.
+        setattr(
+            self.override.__func__,  # Attributes of a method are immutable, so need to use __func__.
+            "__signature__",
+            signature(self.override).replace(
+                parameters=[
+                    Parameter(name="self", kind=Parameter.POSITIONAL_ONLY),
+                    *[
+                        Parameter(
+                            name=k,
+                            annotation=type(default_value),
+                            kind=Parameter.KEYWORD_ONLY,
+                        )
+                        for k in self._registered_options
+                    ],
+                ]
+            ),
+        )
+        # Update docstring for `SettingsManager.override` as well.
+        insert_index = self.override.__doc__.find("\n        Yields")
```

Comment:
how does that behave if you register more than one option?

Reply:
What do you mean?  This looks for the spot right before "Yields" which is where you'd want to insert a new option (i.e., end of the list after "Parameters").  It should work, I checked it on the build with some test options.

Reply:
This code gets called every time a setting is registered.

And every time, `options` gets inserted after `Yields`. Won’t that insert all options multiple times?

Reply:
The next line only adds the new option.  Will relabel the variable.

Reply:
![Screenshot 2024-01-25 at 11 02 03](https://github.com/scverse/anndata/assets/43999641/61da496f-7fd4-4b9f-a3a1-3fbec7f53520)